### PR TITLE
Allow to create incremental password protected backups

### PR DESCRIPTION
**Pull request by sguerrini97**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduced `use_same_password_for_base_backup` settings for `BACKUP` and `RESTORE` queries, allowing to create and restore incremental backups to/from password protected archives.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Review by antaljanosbenjamin — COMMENTED** [review]
The changes look good, two comments:
1. Please revert the change in the `orc` submodule, I think it is an accident
2. As discussed on [community slack](https://clickhousedb.slack.com/archives/CU478UEQZ/p1720433225935729?thread_ts=1720184064.867039&cid=CU478UEQZ), please add some tests.

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit fc15c037b3eba402fd51e72eedd6ae27fafe171b with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66214/fc15c037b3eba402fd51e72eedd6ae27fafe171b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66214/fc15c037b3eba402fd51e72eedd6ae27fafe171b/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by antaljanosbenjamin** [comment]
So check this [comment](https://github.com/ClickHouse/ClickHouse/pull/66214#issuecomment-2213643267) of @robot-ch-test-poll2 . The second line is `Click here...`, click there and you will see the [CI report](https://s3.amazonaws.com/clickhouse-test-reports/66214/1e123fe9839bb1986fdce68dec823eebac60c015/ci_running.html). In the report you can sort the lines based on columns by clicking on the columns, you cansee the short summary of tests by clicking on their name, or open their full report by clicking on link in the `Logs` column.

Once you are seeing the setailed report, you can also click on test names to see the output and you can also download various things (among the logs) by clicking on the items above the table.

File: tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  DROP TABLE IF EXISTS data;
+  DROP TABLE IF EXISTS data_1;
+  DROP TABLE IF EXISTS data_2;
+  CREATE TABLE data (key Int) ENGINE=MergeTree() ORDER BY tuple();
+  INSERT INTO data SELECT * from numbers(10);
+"
+
+echo 'use_same_password_for_base_backup'
+echo "base"
+$CLICKHOUSE_CLIENT -q "BACKUP TABLE data TO Disk('backups', '${CLICKHOUSE_TEST_NAME}_base.zip') SETTINGS password='password';" | cut -f2
```
[the remark this answers is not in the document]
Comment (sguerrini97) [reply tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh:18]: Won't including `_default` postfix fail in other scenarios? E.g. when the test is executed locally.

[the remark this answers is not in the document]
Comment (sguerrini97) [reply tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh:18]: Perhaps I can just grep for `_inc_1.zip': Password is required. (CANNOT_UNPACK_ARCHIVE)` ?

[the remark this answers is not in the document]
Comment (antaljanosbenjamin) [reply tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh]: > Won't including _default postfix fail in other scenarios? E.g. when the test is executed locally.

It should work assuming you are using `clickhouse-test`. `clickhouse-test` is the way to run tests, so it should be fine.

> Perhaps I can just grep for _inc_1.zip': Password is required. (CANNOT_UNPACK_ARCHIVE) ?

I prefer to stricter tests. There might be issues that we do not even think about and a less strict test won't catch it.

[the remark this answers is not in the document]
Comment (sguerrini97) [reply tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh]: > It should work assuming you are using `clickhouse-test`. `clickhouse-test` is the way to run tests, so it should be fine.

Ok, I understand.
Sorry but it's my first time digging into the ClickHouse test environment.

> I prefer to stricter tests. There might be issues that we do not even think about and a less strict test won't catch it.

I agree. My last commit should be good :smile:

[the remark this answers is not in the document]
Comment (antaljanosbenjamin) [reply tests/queries/0_stateless/02843_backup_use_same_password_for_base_backup.sh]: > Sorry but it's my first time digging into the ClickHouse test environment.

Don't worry, it is a big ecosystem, everybody need some time to get familiar with it. If you are planning to contribute also in the future, you can read our [contribution guide](https://clickhouse.com/docs/en/development/developer-instruction) (in the left side menu you can scroll down) that includes how do we deal with [tests](https://clickhouse.com/docs/en/development/tests) too.

**Comment by sguerrini97** [comment]
@antaljanosbenjamin hi, do you mind checking the failed checks?
I don't see any connection with my code. Thanks!

**Comment by antaljanosbenjamin** [comment]
I will do check them later today or tomorrow.

**Comment by antaljanosbenjamin** [comment]
- `AST fuzzer (debug)` https://github.com/ClickHouse/ClickHouse/pull/66395
 - `Performance Comparison (release) [4/4]` https://github.com/ClickHouse/ClickHouse/pull/66342

The other failures are trickier. Trying to figure out what happened there.

**Comment by antaljanosbenjamin** [comment]
@sguerrini97 could please merge a master? There were some issues in our CI and I think it is better to rerun the tests to make sure your PR doesn't break anything.

**Comment by antaljanosbenjamin** [comment]
Do you only merge master because of failed tests? If so, please stop it. We have some flaky tests, I have to check them. I check the PRs once or twice a day, and when I check you PR it was always running CI, thus I couldn't check the tests.

**Comment by antaljanosbenjamin** [comment]
No, it is not. I just have to check out the failed integration test and after that I think it could be merged. I know it is stretched quite far, but our CI is in quite a bad shape nowadays, constant fighting. Please do not merge master, this is an exceptionally good CI run, once we can easily merge PRs, I will take a look at this.

**Comment by antaljanosbenjamin** [comment]
So the integration tests seems to be a bit more complicated to understand than usually, but I think here is what happened. So on `node2` there is a session expired exception ~12 seconds before the failure:
```
2024.07.17 15:20:17.973975 [ 2299 ] {} <Information> DDLWorker: Lost ZooKeeper connection, will try to connect again: Code: 999. Coordination::Exception: Session expired. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):
And it seems to start the backup process at 15:20:27:
2024.07.17 15:20:27.918348 [ 2299 ] {} <Debug> DDLWorker: Executing query: RESTORE TABLE system.users FROM Disk('backups', '28') SETTINGS restore_uuid = '8fd92570-0ed1-4fbc-b81f-eda9279931f4', internal = true, async = true, host_id = 'node2:9000', cluster_host_ids = [['node1:9000', 'node2:9000']]
```
while `node1` declares the backup failed at `15:20:23`:
2024.07.17 15:20:23.771985 [ 9 ] {1922533f-42f5-4474-90bf-383f0de2193f} <Error> BackupsWorker: Failed to restore from backup Disk('backups', '28'): Code: 655. DB::Exception: No connection to host node2:9000 yet. (FAILED_TO_SYNC_BACKUP_OR_RESTORE), Stack trace (when copying this message, always include the lines below):
```

Is it possible that the backup failed just because the reconnection to keeper took a few seconds? I think yes, but I have to confirm it.

Also on `node1` there are strange errors right before failing the backup:
```
2024.07.17 15:20:23.685164 [ 682 ] {08eec6da-0d20-46b1-93fe-694aa271c906} <Trace> RestoreCoordinationRemote: ZooKeeperRetriesControl: waitImpl: setUserError: error=655 message=No connection to host node1:9000 yet, will retry
2024.07.17 15:20:23.685252 [ 682 ] {08eec6da-0d20-46b1-93fe-694aa271c906} <Debug> RestoreCoordinationRemote: ZooKeeperRetriesControl: waitImpl: will retry due to error: retry_count=992/1000 timeout=1ms error=655 message=No connection to host node1:9000 yet, will retry
2024.07.17 15:20:23.753371 [ 682 ] {} <Error> BackupsWorker: Failed to restore from internal backup Disk('backups', '28'): Code: 999. Coordination::Exception: Coordination error: No node, path /clickhouse/backups/restore-8fd92570-0ed1-4fbc-b81f-eda9279931f4/stage. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):
```
But I think this can caused by cleaning up the paths in keeper once the backup is considered failed. However I couldn't confirm that this is the case.

**Comment by vitlibar** [comment]
@antaljanosbenjamin You're right, the reason of this test failure was that `node2` waited for 10 seconds (`15:20:17 - 15:20:27`) before it could reconnect to ZooKeeper, while `node1` waited only 1 second (`15:20:22 - 15:20:23`) for a response from `node2` (via ZooKeeper). So at `15:20:23` `node1` decided to stop waiting and consider the whole RESTORE failed. I'll increase timeouts in this test. This test failure isn't related to this PR's changes.

**Pull request merged**